src/NUnitTestAdapter/TraitsFeature.cs

```diff
@@ -59,7 +59,45 @@ public sealed class CachedTestCaseInfo
             // an immutable struct. (https://github.com/nunit/nunit3-vs-adapter/pull/457)
         }
 
-        public static void AddTraitsFromTestNode(this TestCase testCase, NUnitTestCase testNCase,
+        public static void AddTraitsFromTestNode(this TestCase testCase, NUnitDiscoveryTestCase testNCase,
+            IDictionary<string, CachedTestCaseInfo> traitsCache, ITestLogger logger, IAdapterSettings adapterSettings)
+        {
+            var ancestor = testNCase.Parent as INUnitDiscoverySuiteBase;
+            var key = ancestor?.Id;
+            var categoryList = new CategoryList(testCase, adapterSettings);
+            // Reading ancestor properties of a test-case node. And adding to the cache.
+            while (ancestor != null && !string.IsNullOrEmpty(key))
+            {
+                if (traitsCache.ContainsKey(key))
+                {
+                    categoryList.AddRange(traitsCache[key].Traits.Where(o => o.Name == NunitTestCategoryLabel).Select(prop => prop.Value).ToList());
+
+                    if (traitsCache[key].Explicit)
+                        testCase.SetPropertyValue(CategoryList.NUnitExplicitProperty, true);
+
+                    var traitsList = traitsCache[key].Traits.Where(o => o.Name != NunitTestCategoryLabel).ToList();
+                    if (traitsList.Count > 0)
+                        testCase.Traits.AddRange(traitsList);
+                }
+                else
+                {
+                    categoryList.ProcessTestCaseProperties(ancestor, true, key, traitsCache);
+                    // Adding entry to dictionary, so that we will not make SelectNodes call again.
+                    if (categoryList.LastNodeListCount == 0 && !traitsCache.ContainsKey(key))
+                    {
+                        traitsCache[key] = new CachedTestCaseInfo();
+                    }
+                }
+                ancestor = ancestor.Parent;
+                key = ancestor?.Id;
+            }
+
+            // No Need to store test-case properties in cache.
+            categoryList.ProcessTestCaseProperties(testNCase, false);
+            categoryList.UpdateCategoriesToVs();
+        }
```

Comment:
Could all the commonalities be extracted to a helper method with the necessary values passed in?

Reply:
Seems like a good idea.  I haven't touched this now, but if we have enough tests around it, I'll give it a run.

Reply:
Postpones to later